determine_missing_jobs finds both opt and pgo jobs if available

mandaltapesh opened the pull request:
Hello armenzg,

made some changes . It passes all the test.Maybe because the change is not effective.Kindly see.

r? armenzg

Regards,
Tapesh Mandal

armenzg commented:
Hi @mandaltapesh I'm glad to see that you proposed a change.

In the issue I gave you a piece of code that would help you make sure that your changes would work.
Here's a more precise example that should not return both True [1]

Please make sure that this simple test works for you before the next review request.

A trick you might want to use when testing your code is to insert this line to do debugging:
> import pdb; pdb.set_trace()
You can use it to print the values of the variable names and even test conditional statements.

[1]
```python
In [1]: from mozci.platforms import _wanted_builder

In [2]: _wanted_builder('Windows XP 32-bit mozilla-inbound pgo test xpcshell')
Out[2]: True

In [3]: _wanted_builder('Windows XP 32-bit mozilla-inbound opt test xpcshell')
Out[3]: True
```

mandaltapesh commented:
Hi armenzg,
Made some changes . Tested them as you mentioned above.

(venv)tapesh@tapesh-HP-240-G4-Notebook-PC:~/Desktop/mozilla_ci_tools$ python -c "from mozci.platforms import _wanted_builder;x = _wanted_builder('Windows XP 32-bit mozilla-inbound pgo test xpcshell');print x"
False
(venv)tapesh@tapesh-HP-240-G4-Notebook-PC:~/Desktop/mozilla_ci_tools$ python -c "from mozci.platforms import _wanted_builder;x = _wanted_builder('Windows XP 32-bit mozilla-inbound opt test xpcshell');print x"
True

also attaching a screenshot :)

![mozilla](https://cloud.githubusercontent.com/assets/6252431/13462091/921914ea-e0ab-11e5-9157-e279ba5a69c3.png)

r? armenzg

Regards
Tapesh Mandal

armenzg commented:
Well done!

Now for mozilla-aurora, mozilla-beta, mozilla-release and other release repositories the story is a bit different.

The 'opt' builders exists in Buildbot (which is in reality a bug), however, they are not scheduled for those repositories. [1]

Could you please adjust your PR to accomodate for these exceptions? Buildbot scheduling has always been this messy.

We will need tests for this.
You will need a new class in here:
https://github.com/mozilla/mozilla_ci_tools/blob/master/test/test_platforms.py

[1] https://gist.github.com/armenzg/0bf88306e35bdddbbaec

mandaltapesh commented:
hey armenzg 
Was busy today...made some changes to keep the wheels turning...
added tests in a new class with two functions , modified mock_all_things, and took the buildbot related exception into consideration and modified the "if" statement.

Please correct me wherever i go wrong..i was also wondering if you give me specific testcases for the buildbot exceptions so that i can verify it inline by calling the _wanted_builder(test_case) ....maybe its there in the text file you gave https://gist.github.com/armenzg/0bf88306e35bdddbbaec but can you please pinpoint it .

So far the changes i made ran without any errors with all tests passing and no flake errors .Checked that using tox under venv.

Have a happy weekend :)

r? armenzg

Regards,
Tapesh Mandal

armenzg commented:
@tapeshmandal part of software development is to read requirements and trying to understand what is going on. If I do point things to you I'm taking away from you doing the mental effort. In fact, the hardest and most time consuming part is to get the requirements right. If I do it all, you want be learning as much.

I will give you one last tip and that's it; Run the same things as in https://github.com/mozilla/mozilla_ci_tools/pull/441#issuecomment-191245080
but with the builders below [1]

I don't want you to give me the results you get; I want to tell me of those results which ones should be different.
Read this comment to know which repositories should be different:
> Now for mozilla-aurora, mozilla-beta, mozilla-release and other release repositories the story is a bit different.

These are the case scenarios:
* 1 builder on that repo
* We want 1 builder out of 2 builders on that repo (not mozilla-aurora et all)
* We want 2 builders out of 2 builders on that repo (remaining repositories)

[1]
 u'TB Windows XP 32-bit comm-aurora opt test xpcshell',
 u'TB Windows XP 32-bit comm-beta opt test xpcshell',
 u'TB Windows XP 32-bit comm-central opt test xpcshell',
 u'TB Windows XP 32-bit comm-esr38 opt test xpcshell',
 u'TB Windows XP 32-bit comm-esr45 opt test xpcshell',
 u'TB Windows XP 32-bit cypress opt test xpcshell',
 u'TB Windows XP 32-bit try-comm-central opt test xpcshell',
 u'Windows XP 32-bit ash opt test xpcshell',
 u'Windows XP 32-bit ash pgo test xpcshell',
 u'Windows XP 32-bit date opt test xpcshell',
 u'Windows XP 32-bit date pgo test xpcshell',
 u'Windows XP 32-bit fx-team opt test xpcshell',
 u'Windows XP 32-bit fx-team pgo test xpcshell',
 u'Windows XP 32-bit jamun opt test xpcshell',
 u'Windows XP 32-bit mozilla-aurora opt test xpcshell',
 u'Windows XP 32-bit mozilla-aurora pgo test xpcshell',
 u'Windows XP 32-bit mozilla-beta opt test xpcshell',
 u'Windows XP 32-bit mozilla-beta pgo test xpcshell',
 u'Windows XP 32-bit mozilla-central opt test xpcshell',
 u'Windows XP 32-bit mozilla-central pgo test xpcshell',
 u'Windows XP 32-bit mozilla-esr38 opt test xpcshell',
 u'Windows XP 32-bit mozilla-esr38 pgo test xpcshell',
 u'Windows XP 32-bit mozilla-esr45 opt test xpcshell',
 u'Windows XP 32-bit mozilla-esr45 pgo test xpcshell',
 u'Windows XP 32-bit mozilla-inbound opt test xpcshell',
 u'Windows XP 32-bit mozilla-inbound pgo test xpcshell',
 u'Windows XP 32-bit mozilla-release opt test xpcshell',
 u'Windows XP 32-bit mozilla-release pgo test xpcshell',
 u'Windows XP 32-bit oak opt test xpcshell',
 u'Windows XP 32-bit oak pgo test xpcshell',
 u'Windows XP 32-bit try opt test xpcshell',

mandaltapesh commented:
@armenzg   i think i got it 2 out of 2 repos for mozilla-aurora,mozilla-beta,mozilla-esr38,45,mozilla-release,mozilla-inbound,mozilla-central ..i mean to say it should return True for both opt and pgo jobs for these repos.

and for oak,try,ash,date et al it should return False for pgo and true for opt 

Regards,
Tapesh Mandal

armenzg commented:
You won't know for sure unless you write tests for the following repos:
- m-i
- m-c
- m-a
- m-b
- m-r
- m-esr*

From looking at the code I know it won't work.

mandaltapesh commented:
Made the changes as required :)
Thanks for support

Regards,
Tapesh Mandal

armenzg commented on test/mock_allthethings.json:
```diff
@@ -95,6 +109,173 @@
+        "shortname": "mozilla-inbound-platform1",
+        "slavebuilddir": "branch-platform-00000000000000000",
+        "slavepool": "2e490369b232e800931337b9b1186f5027eac1e6"
+    },"Platform1 mozilla-release pgo test mochitest-1": {
```
Could this be moved to the next line?

armenzg commented on test/test_platforms.py:
```diff
@@ -173,6 +174,69 @@ def test_with_build_job(self, fetch_allthethings_data):
+        fetch_allthethings_data.return_value = MOCK_ALLTHETHINGS
+        self.assertEquals(
+            _wanted_builder('Platform1 mozilla-central pgo test mochitest-1'),
+            True)
```
This should be False.

Look at what I wrote in the gist:

> * mozilla-inbound --> opt [1] (every few hours PGO builds are scheduled)
> * mozilla-central --> opt [2] (every few hours PGO builds are scheduled)
> * mozilla-aurora --> pgo [3]
> * mozilla-beta --> pgo [4]
> * mozilla-release --> pgo [4]
> * mozilla-esr{38,45} --> pgo [5]

armenzg commented on mozci/platforms.py:
```diff
@@ -495,6 +495,22 @@ def _wanted_builder(builder, filter=True, repo_name=None):
+            'mozilla-esr45',
+            'mozilla-inbound',
+            'mozilla-release') and \
+           info['job_type'] == 'test' and _get_job_type(builder) == 'pgo':
```
You will need to adjust this to make the tests pass.
Also add a note this note in here:

> On Buildbot, many repos have both opt and pgo builders, however, when it comes
> to adding missing jobs to a push we're generally interested on the opt builder since that is what 
> matters to sheriffs.
>
> Unfortunately, the list of builders on Buildbot is a bit of a mess. There might be both opt and
> pgo builders, however, some will *never* be scheduled. For instance, release repos have both
> opt and pgo builders, however, only the pgo builders are scheduled on push.
>
> Here's a table of what we want:
> * repositories with only opt builders; return True
> * mozilla-inbound,fx-team,try --> opt
> * mozilla-central --> opt
> * mozilla-{aurora,beta,release} --> pgo
> * mozilla-esr{38,45} --> pgo

armenzg commented:
This is finally going in the right direction. Please look at the comments very closely and adjust the tests appropriately.

mandaltapesh commented:
@armenzg  Made the above mentioned changes . Please have a look when you are free :)

Regards,
Tapesh Mandal

armenzg commented on test/test_platforms.py:
```diff
@@ -173,6 +174,69 @@ def test_with_build_job(self, fetch_allthethings_data):
+            True)
+        self.assertEquals(
+            _wanted_builder('Platform1 repo pgo test mochitest-1'),
+            False)
```
Now that I think of it we need two more cases.

A new repo with only opt.
Another new repo with only pgo.

It will probably raise an exception when we try to check on a builder that does not exist on MOCK_ALLTHETHINGS.
Once _wanted_builder() calls get_buildername_metadata() with an invalid builder we should get an exception [1].
Could you please raise a MozciError here instead of None?

[1] https://github.com/mozilla/mozilla_ci_tools/blob/master/mozci/platforms.py#L478
[2] https://github.com/mozilla/mozilla_ci_tools/blob/master/mozci/platforms.py#L222

armenzg replied:
Filed as https://github.com/mozilla/mozilla_ci_tools/issues/442

armenzg commented:
Congratulations! and thanks for your help!

mandaltapesh commented:
you are welcome :)
